File: backend/Model/Quotation.cs
using System;

namespace YLWorks.Model
{
    public class Quotation: BaseEntity
    {
        public Guid Id {  get; set; }
        public string? QuotationNo { get; set; } = string.Empty;
        public DateTime QuotationDate { get; set; }
        public Guid FromCompanyId { get; set; }
        public Company FromCompany { get; set; }
        public Guid ClientId { get; set; }
        public Company Client { get; set; } = null!;
        public string? ProjectCode { get; set; }
        public Project? Project { get; set; }
        public string? Subject { get; set; }

        public decimal? SubTotal { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TaxAmount { get; set; }
        public decimal? TotalAmount { get; set; }

        public string? PaymentTerms { get; set; }
        public int? Validity { get; set; }
        public string? ValidityType { get; set; }

        public DateTime? DueDate { get; set; }

        public string Status { get; set; } = "Draft"; // Draft, Revised, Approved, Sent, Accepted, Rejected
        public string? Remarks { get; set; }
        public Guid CreatedById { get; set; }
        public User CreatedBy { get; set; } = null!;
        public ICollection<QuotationStatusHistory> QuotationStatusHistories { get; set; } = new List<QuotationStatusHistory>();
        public ICollection<QuotationItems> QuotationItems { get; set; } = new List<QuotationItems>();
        public ICollection<QuotationTermsAndCondition> TermsAndConditions { get; set; }
    = new List<QuotationTermsAndCondition>();
        public ICollection<QuotationOtherInformation> QuotationOtherInformations { get; set; }
    = new List<QuotationOtherInformation>();
    }

    public class QuotationStatusHistory
    {
        public Guid Id { get; set; }
        public Guid QuotationId { get; set; }
        public Quotation Quotation { get; set; } = null!;
        public string Status { get; set; } = string.Empty;
        public DateTime ActionAt { get; set; }
        public Guid? ActionUserId { get; set; }
        public User? ActionUser { get; set; }
        public string? Remarks { get; set; }
        public string? SignatureImage { get; set; }
    }

    public class QuotationItems : BaseEntity
    {
        public Guid Id { get; set; }
        public Guid QuotationId { get; set; }
        public Quotation Quotation { get; set; } = null!;
        
        public Guid? ProductServiceId { get; set; }
        public ProductService? ProductService { get; set; }

        public string? RowType { get; set; } // CategoryHeader, LineItem, NoteRow
        public string? Item { get; set; }
        public string? Description { get; set; }
        public decimal? Quantity { get; set; }
        public string? Unit { get; set; }
        public decimal? UnitPrice { get; set; }
        public decimal? Discount {  get; set; }
        public decimal? TotalPrice { get; set; }
        public int SortOrder { get; set; }
    }

    public class QuotationTermsAndCondition
    {
        public Guid Id { get; set; }
        public Guid QuotationId { get; set; }
        public Quotation Quotation { get; set; }
        public Guid TermsAndConditionId { get; set; }
        public TermsAndCondition TermsAndCondition { get; set; }
        public int SortOrder { get; set; }
    }

    public class QuotationItemDto
    {
        public Guid Id { get; set; }
        public Guid QuotationId { get; set; }
        public Guid? ProductServiceId { get; set; }
        public string? RowType { get; set; } = "LineItem";
        public string? Item { get; set; }
        public string? Description { get; set; }
        public decimal? Quantity { get; set; }
        public string? Unit { get; set; }
        public decimal? UnitPrice { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalPrice { get; set; }
        public int SortOrder { get; set; }
    }

    public class QuotationItemBase
    {
        public Guid? Id { get; set; }
        public Guid QuotationId { get; set; }

        public Guid? ProductServiceId { get; set; } 

        public string? RowType { get; set; } = "LineItem";
        public string? Item { get; set; }
        public string? Description { get; set; }
        public decimal? Quantity { get; set; }
        public string? Unit { get; set; }
        public decimal? UnitPrice { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalPrice { get; set; }
        public int SortOrder { get; set; }
    }

    public class QuotationItemRequest : QuotationItemBase 
    {
        public List<QuotationItemRequest> Children { get; set; } = new();
    }

    public class UpdateQuotationItemRequest : QuotationItemBase
    {
        public List<QuotationItemRequest> Children { get; set; } = new();
    }

    public abstract class BaseQuotationRequest
    {
        public string? QuotationNo { get; set; } = string.Empty;
        public DateTime QuotationDate { get; set; }
        public Guid FromCompanyId { get; set; }
        public Guid ClientId { get; set; }
        public string? ProjectCode { get; set; }
        public string? Subject { get; set; }
        public decimal? SubTotal { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TaxAmount { get; set; }
        public decimal TotalAmount { get; set; }
        public string? PaymentTerms { get; set; }
        public int? Validity { get; set; }
        public string? ValidityType { get; set; }
        public List<TermsAndConditionOrderDto> TermsAndConditions { get; set; } = new();
        public List<QuotationOtherInfoRequest> QuotationOtherInformations { get; set; }
        = new();
    }

    public class CreateQuotationRequest : BaseQuotationRequest
    {
        public List<QuotationItemRequest> QuotationItems { get; set; } = new();
    }

    public class UpdateQuotationRequest : BaseQuotationRequest
    {
        public Guid Id { get; set; }
        public List<UpdateQuotationItemRequest> QuotationItems { get; set; } = new();
    }

    public class UpdateQuotationStatusRequest
    {
        public Guid Id { get; set; }          
        public string? Status { get; set; }  
        public string? Remarks { get; set; }  
        public string? SignatureImage { get; set; }
    }

    public class QuotationDropdownDto
    {
        public Guid Id { get; set; }
        public string QuotationNo { get; set; } = string.Empty;
        public decimal? TotalAmount { get; set; }
        public Guid? ClientId { get; set; }
        public Guid? Project { get; set; }
        public string? ProjectCode { get; set; }
        public string? ProjectTitle { get; set; }
        public Guid? FromCompanyId {  get; set; }
        public string? CompanyName { get; set; }
        public List<QuotationItemDto>? Items { get; set; } = new();

    }

    public class ConvertQuotationToSoRequest
    {
        public string? ClientPONumber { get; set; }
        public DateTime? ClientPODate { get; set; }
        public IFormFile? ClientPOAttachment { get; set; }
        public string? Remarks { get; set; }
    }

    public class QuotationOtherInformation
    {
        public Guid Id { get; set; }

        public Guid QuotationId { get; set; }
        public Quotation Quotation { get; set; } = null!;

        public string Key { get; set; } = string.Empty;     
        public string Value { get; set; } = string.Empty;   

        public int SortOrder { get; set; }
    }

    public class QuotationOtherInfoRequest
    {
        public string Key { get; set; } = string.Empty;
        public string Value { get; set; } = string.Empty;
    }

    public class TermsAndConditionOrderDto
    {
        public Guid? Id { get; set; } 
        public string? Title { get; set; }             
        public string? Description { get; set; }      
        public int SortOrder { get; set; }
    }

}

File: frontend/src/app/models/Quotation.ts
import { BaseModel } from './BaseModel';
import { CompanyDto } from './Company';
import { ProductServiceDto } from './ProductService';
import { ProjectDto } from './Project';
import { TermsAndConditionDto } from './TermsAndCondition';
import { UserDto } from './User';

export interface QuotationDto extends BaseModel {
  quotationNo: string;
  quotationDate: Date;
  fromCompanyId: string;
  fromCompany: CompanyDto;
  clientId: string;
  client: CompanyDto;
  projectCode: string;
  project: ProjectDto;
  subject: string;
  discount: number;
  subTotal: number;
  taxAmount: number;
  totalAmount: number;
  paymentTerms: string;
  validity: number;
  validityType: string;
  status: string;
  remarks: string;
  quotationItems: QuotationItems[];
  quotationStatusHistories: QuotationStatusHistory[];
  termsAndConditions: QuotationTermsAndCondition[];
  otherInformations: QuotationOtherInformation[];
  createdBy: UserDto;
}

export interface QuotationStatusHistory extends BaseModel {
  quotationId: string;
  quotation: QuotationDto;
  status: string;
  actionAt: Date;
  actionUserId: string;
  remarks: string;
  actionUser?: {
    id: string;
    fullName: string;
  };
}

export interface QuotationItems extends BaseModel {
  quotationId: string;
  productServiceId?: string;
  productService?: ProductServiceDto;
  rowType: string;
  item: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  totalPrice: number;
  sortOrder: number;
}

export interface QuotationTermsAndCondition {
  id: string;
  quotationId: string;
  termsAndConditionId: string;
  termsAndCondition: TermsAndConditionDto;
  sortOrder: number;
}

export interface QuotationItemDto {
  id: string;
  quotationId: string;
  productServiceId?: string;
  rowType: string;
  item: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  totalPrice: number;
  sortOrder: number;
}

export interface QuotationItemBase {
  id?: string | null;
  quotationId?: string;

  productServiceId?: string | null;

  rowType: string;
  item: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  totalPrice: number;
  sortOrder: number;
}

export interface QuotationItemRequest extends QuotationItemBase {}

export interface UpdateQuotationItemRequest extends QuotationItemBase {
  id: string;
}

export interface CreateQuotationRequest {
  quotationNo: string;
  quotationDate: Date;
  fromCompanyId: string;
  clientId: string;
  projectCode: string;
  subject: string;

  taxAmount: number;
  discount: number;
  subTotal: number;
  totalAmount: number;

  paymentTerms: string;
  validity: number;
  validityType: string;

  quotationItems: QuotationItemRequest[];

  termsAndConditions: TermsAndConditionOrderDto[];

  otherInformations: QuotationOtherInformation[];
}

export interface UpdateQuotationRequest extends CreateQuotationRequest {
  id: string;
  quotationItems: UpdateQuotationItemRequest[];
}

export interface UpdateQuotationStatusRequest {
  id: string;
  status: string;
  signatureImage: string;
}

export interface QuotationOtherInformation {
  id?: string;
  quotationId?: string;
  key: string;
  value: string;
  sortOrder?: number;
}

export interface TermsAndConditionOrderDto {
  termsAndConditionId?: string | null;
  title?: string;
  description?: string;
  sortOrder: number;
}
